### DISCUSSION TITLE
How to create validated records when using custom models?

### DISCUSSION THREAD

**Discussion by lumosminima** [pocketbase/pocketbase#4724] [Q&A] [answered]
👋 [This example](https://pocketbase.io/docs/go-custom-models#create-article) on custom model docs page only shows saving data without validations but what about validated data which is shown [here](https://pocketbase.io/docs/go-records/#create-new-record-with-data-validations) using `forms.NewRecordUpsert` I tried passing my custom model struct like `forms.NewRecordUpsert(pb,customModel)` but get compiler error:
```
customModel (type) is not an expression compiler (NotAnExpr)
```
I'm new to Go coming from typescript so don't know what exactly this error means but I reverted to using the docs way:
```GO
        collection, err := app.Dao().FindCollectionByNameOrId("myCollection")

	if err != nil {
		return nil, err
	}

        record := models.NewRecord(collection)

	form := forms.NewRecordUpsert(app, record)
```
But now my return type doesn't match `*customModel` and needs to be `*models.Record` My data will need to depend on other tables and custom recursive CTE for the responses which is why I need to extend which bring me to my second question: Do I even need a custom model? what use cases they allow? because I didn't get any type intellisense for my db columns on `app.Dao()` and `dbx` methods which I assume it would. Previously, I was just using simple structs like: `ProjectRequest`,`ProjectResponse`,`Project` which had different struct fields based on service layer which I was using inside my handlers which were simple functions that had this signature:
```GO
type Project struct {
	...
}

type ProjectRequest struct {
	...
}

type ProjectResponse struct {
	...
}

// handler for echo
func HandleCreateProject(app *pocketbase.PocketBase, c echo.Context) error {
    requestData  := new(ProjectRequest)

	if err := c.Bind(requestData); err != nil {
		return err
	}
      
     newProject := &Project {
             ...requestData  
     }

     newProjectResponse, err := CreateNewProject(app, newProject)

	if err != nil {
		return err
	}

	return c.JSON(http.StatusOK, newProjectResponse)
}
// calls this inside which returns ProjectResponse which has some additional fields for clients
func CreateNewProject(app *pocketbase.PocketBase, project *ProjectRequest) (*models.Record, error) 
```
This allowed some type safety but it is manual and I will need to keep updating these structs to match my db fields vs only updating the transformed structs like RequestData and ResponseData and having one struct that matches my db columns coming from pocketbase. But before that comes is this pattern acceptable in GO or are there better more idiomatic ways to handle it? I also saw that upcoming refactoring will probably solve the validation part with a single record interface on app but should I keep using the custom model (will there be enhancements to that workflow) or just use the simpler but generic *models.Record and its helpers?

_P.S thank you for this awesome project 💖 It allowed me to get into learning GO and I'm having lots of fun!_

**Comment by pnmcosta** [comment]
Fyi, this is most likely changing in near future, see https://github.com/pocketbase/pocketbase/discussions/4355 (i believe `forms` are being dropped and replaced with a new way to custom model and DaO interfaces).

I'm also summarizing from my experience:

- I have custom models, they implement `Validatable` interface which ensures the model has `Validate() error` method implemented, I have overloads for dao Save records that ensure validation occurs. It is mostly driven by [this package](github.com/go-ozzo/ozzo-validation) however I only really need it for custom validations.
- I only use `forms.NewRecordUpsert` for auth collections, because of PB's own rules on creating auth records.
- If I ever need to use both Records and Custom models I ensure the db and json tags match, that way I can use a simple unmarshal/marshal to cast from a record to a model, e.g.:

```
import (
	"encoding/json"
	"fmt"

	pb "github.com/pocketbase/pocketbase/models"
)

// Note: db and json field must match because of pb.Record being leveraged
type Account struct {
	pb.BaseModel

	Email        string         `db:"email" json:"email"`
	CountryCode  string         `db:"countryCode" json:"countryCode"`
}

func MapRecordToAccount(record *pb.Record) (*Account, error) {
	data, err := json.Marshal(record)
	if err != nil {
		return nil, fmt.Errorf("marshal account record error: %v", err)
	}

	account := Account{}
	err = json.Unmarshal(data, &account)
	if err != nil {
		return nil, fmt.Errorf("unmarshal account record error: %v", err)
	}

	return &account, nil
}
```

**Reply by lumosminima** [reply]
I get an error when using the mapping function:
```
json: cannot unmarshal array into Go struct field Account.type of type string
```
My custom model is same and the `record *pb.Record` that I pass is the validated record that is created via `forms.NewRecordUpsert`. But also wouldn't I have to create such a function for each custom model which seems not very DRY. I also don't see any db columns name suggestion on Dao() or dbx when using custom model so don't know what is the actual benefit in using them vs the basic `*models.Record` apart from a more concrete return type: `*Account` instead of `*models.Record`

P.S if all of this will be solved by #4355 then should I just use `*models.Record` for now?

**Reply by ganigeorgiev** [reply, marked as answer]
@lumosminima If you are not sure that you need custom struct models, prefer using directly `*models.Record` as it comes with a lot less caveats.

The custom model structs were initially intended to allow interacting directly with the database, for example scanning the rows of a custom raw SQL queries that returns columns from multiple tables.

With the refactoring custom models as it is right now will be discouraged and instead wrappers like the ones suggested in https://github.com/pocketbase/pocketbase/discussions/2697#discussioncomment-6154067 will be the recommended way to define a typed struct for your models (_eventually there will be an option to generate them via console command_).

**Reply by lumosminima** [reply]
@ganigeorgiev Great! though proper typing for custom raw queries was one of the reason I started looking into custom models like I have a lot of queries like this which touch multiple tables to get balance and such:
```
queryError := app.Dao().DB().
		NewQuery(CTEQueryString).Bind(params).
		All(&accounts)
```
This would also be a lot simpler if view collection allowed params! I'll wait for the refactoring 🤞 and move to `*models.Record` for now.

**Reply by pnmcosta** [reply]
@lumosminima seems like a type mapping issue, your `type` field on the custom model is set as type string, but the record's field is actually a slice/array.

But since you're sticking with `models.Record` you won't need to helper above.
